chainercv/utils/iterator/apply_to_batch.py

```diff
@@ -0,0 +1,153 @@
+from chainercv.utils.iterator.unzip import unzip
+
+
+def apply_to_batch(func, iterator, n_input=1, hook=None):
+    """Apply a function/method to an iterator of batches.
+
+    This function applies a function/method to an iterator of batches.
+    It assumes that the iterator returns a batch of input data or
+    a batch of tuples whose first elements ara input data.
```

Comment:
I thought it was better to describe the internals of this function more in the beginning.
I wrote an alternative docstring. What do you think? 

---

This function assumes that the iterator iterates over a collection of tuples that contain inputs to `func`.
Additionally, the tuples may contain values that are not used by `func`.
For convenience, we allow the iterator to iterate over a collection of inputs that are not tuple.
t
Conceptually, this function does two operations.
First, it splits the iterator into two iterators. The first iterator contains the inputs to `func` and the second iterator contains the rest of the data from `iterator`.
Second, it creates another iterator that iterates over a collection of the outputs of the `func`.

Here is an illustration of the expected behavior of `iterator`.
`M` is the number of arguments for `func`, which is defined by an argument `n_input`.  `N` is the number of elements in a tuple from `iterator` that are not used by `func`. Note that `N` can be 0.

```
>>> batch = next(iterator)
>>> # batch:  [(in_val0, ..., in_val{M-1}, rest_val0, ..., rest_val{N-1})]
or
>>> # batch:  [in_val]
```

Reply:
> First, it splits the iterator into two iterators. The first iterator contains the inputs to func and the second iterator contains the rest of the data from iterator.
> Second, it creates another iterator that iterates over a collection of the outputs of the func.

This explanation sounds this function returns three iterators. However, it is not correct.

Reply:
Right. That is my mistake.

Reply:
OK, I will try to merge your solution into docstring.